### PR TITLE
[torch.special] Adding betainc

### PR DESCRIPTION
**Pull request by nicola-decao**
Reference: https://github.com/pytorch/pytorch/issues/50345

Adding the incomplete beta function.

TODO
- [x] Bolierplace with a dummy function
- [x] Porting scipy function from https://github.com/scipy/scipy/blob/master/scipy/special/cephes/incbet.c
- [ ] Write test

I am following https://github.com/pytorch/pytorch/pull/55138 as an example of how to implement this. I have an issue: most operation in `torch` are either unary or binary and I cannot find an example of a ternary operation. Right now I have a compilation error:
```
.../pytorch/aten/src/ATen/native/BinaryOps.cpp:61:54: error: too many arguments to function call, expected 3, have 4
  build_binary_float_op(maybe_get_output(), self, a, b);
```
basically, following the implementation of `torch.special.xlog1py` that is:
```
TORCH_META_FUNC(special_xlog1py) (const Tensor& self, const Tensor& other) {
  build_binary_float_op(maybe_get_output(), self, other);
}
```

I have
```
TORCH_META_FUNC(special_betainc) (const Tensor& self, const Tensor& a, const Tensor& b) {
  build_binary_float_op(maybe_get_output(), self, a, b);
}
```
but it is clearly giving an error since it is not binary but ternary. Any suggestions on how to proceed? @kshitij12345

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -713,6 +713,19 @@
 - name: special_xlog1py.other_scalar(Tensor self, Scalar other) -> Tensor
   self: grad * log1p(other.toDouble())
 
+- name: special_betainc(Tensor self, Tensor a, Tensor b) -> Tensor
+  self: grad * 1
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:717]: no need to do * 1 here, you can just return grad as-is if the function has a Jacobian which is an identity matrix (I don't know the function so maybe that's not true?).

Comment (nicola-decao) [reply tools/autograd/derivatives.yaml:717]: Sure. Right now this is a bolierplate. There will be the true grad once implemented.

**Comment by kshitij12345** [comment]
@nicola-decao you are right, we can't directly use the `binary` operator machinery.

AFAIK, you will have to manually add the support for your case (i.e. there is no helper like build_binary_float_op`

Eg snippet to that you can probably use

```c++
TORCH_META_FUNC(special_betainc) (const Tensor& self, const Tensor& a, const Tensor& b) {
    TensorIteratorConfig()
    .set_check_mem_overlap(true)               
    .allow_cpu_scalars(true)                   
    .promote_inputs_to_common_dtype(true)      
    .cast_common_dtype_to_outputs(true)        
    .enforce_safe_casting_to_output(true)      
    .promote_integer_inputs_to_float(true)
    .add_output(maybe_get_output()).
    .add_input(self)
    .add_input(a)
    .add_input(b);
}
```

Reference to `build_binary_float_op`:
https://github.com/pytorch/pytorch/blob/6d97a80dd2f2d9728283d7749b41fdda099417d2/aten/src/ATen/TensorIterator.cpp#L767-L782

Also not sure which file this function should reside. Maybe make a new `Special.cpp`?
@mruberry any suggestions?

**Comment by nicola-decao** [comment]
@kshitij12345 The function will occupy a lot of lines. I would be in favour of a new Special.cpp

**Comment by nicola-decao** [comment]
@kshitij12345 I added what you suggested but I got another error I cannot parse.
```
Undefined symbols for architecture x86_64:
  "at::native::special_betainc(at::Tensor const&, at::Tensor const&, c10::Scalar const&)", referenced from:
      at::compositeexplicitautograd::special_betainc(at::Tensor const&, at::Tensor const&, c10::Scalar const&) in RegisterCompositeExplicitAutograd.cpp.o
      c10::impl::make_boxed_from_unboxed_functor<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&, at::Tensor const&, c10::Scalar const&), &(at::(anonymous namespace)::(anonymous namespace)::wrapper_special_betainc_b_scalar(at::Tensor const&, at::Tensor const&, c10::Scalar const&))>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, at::Tensor const&, c10::Scalar const&> >, false>::call(c10::OperatorKernel*, c10::OperatorHandle const&, c10::DispatchKeySet, std::__1::vector<c10::IValue, std::__1::allocator<c10::IValue> >*) in RegisterCompositeExplicitAutograd.cpp.o
      c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&, at::Tensor const&, c10::Scalar const&), &(at::(anonymous namespace)::(anonymous namespace)::wrapper_special_betainc_b_scalar(at::Tensor const&, at::Tensor const&, c10::Scalar const&))>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, at::Tensor const&, c10::Scalar const&> >, at::Tensor (at::Tensor const&, at::Tensor const&, c10::Scalar const&)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, c10::Scalar const&) in RegisterCompositeExplicitAutograd.cpp.o
  "at::native::special_betainc_out(at::Tensor const&, at::Tensor const&, c10::Scalar const&, at::Tensor&)", referenced from:
      at::compositeexplicitautograd::special_betainc_out(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::Scalar const&) in RegisterCompositeExplicitAutograd.cpp.o
      at::compositeexplicitautograd::special_betainc_outf(at::Tensor const&, at::Tensor const&, c10::Scalar const&, at::Tensor&) in RegisterCompositeExplicitAutograd.cpp.o
      c10::impl::make_boxed_from_unboxed_functor<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor& (at::Tensor const&, at::Tensor const&, c10::Scalar const&, at::Tensor&), &(at::(anonymous namespace)::(anonymous namespace)::wrapper_special_betainc_out_b_scalar(at::Tensor const&, at::Tensor const&, c10::Scalar const&, at::Tensor&))>, at::Tensor&, c10::guts::typelist::typelist<at::Tensor const&, at::Tensor const&, c10::Scalar const&, at::Tensor&> >, false>::call(c10::OperatorKernel*, c10::OperatorHandle const&, c10::DispatchKeySet, std::__1::vector<c10::IValue, std::__1::allocator<c10::IValue> >*) in RegisterCompositeExplicitAutograd.cpp.o
      c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor& (at::Tensor const&, at::Tensor const&, c10::Scalar const&, at::Tensor&), &(at::(anonymous namespace)::(anonymous namespace)::wrapper_special_betainc_out_b_scalar(at::Tensor const&, at::Tensor const&, c10::Scalar const&, at::Tensor&))>, at::Tensor&, c10::guts::typelist::typelist<at::Tensor const&, at::Tensor const&, c10::Scalar const&, at::Tensor&> >, at::Tensor& (at::Tensor const&, at::Tensor const&, c10::Scalar const&, at::Tensor&)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, c10::Scalar const&, at::Tensor&) in RegisterCompositeExplicitAutograd.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/libtorch_cpu.dylib] Error 1
make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by kshitij12345** [comment]
You haven't defined the <Tensor, Tensor, Scalar> variant. Hence the undefined reference.

**Comment by nicola-decao** [comment]
Thanks @kshitij12345 ! Now I fixed some undefined references and it compiles! I can see that `torch.special.betainc` exists but when calling it it gives an error:
```
>>> import torch
>>> torch.special.betainc()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: special_betainc() received an invalid combination of arguments - got (), but expected one of:
 * (Tensor input, Tensor a, Tensor b, *, Tensor out)
 * (Tensor input, Number a, Tensor b, *, Tensor out)
 * (Tensor input, Tensor a, Number b, *, Tensor out)
 * (Tensor input, Number a, Number b, *, Tensor out)

>>> torch.special.betainc(torch.tensor(1.), torch.tensor(1.), torch.tensor(1.))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Unknown device: 16. If you have recently updated the caffe2.proto file to add a new device type, did you forget to update the DeviceTypeName() function to reflect such recent changes?
```

Is there something wrong with the kernel? https://github.com/pytorch/pytorch/blob/7b88409855c6e2ad1af59b39b07bbcc91ee263e8/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L976
Basically I copy-pasted `xlog1py_kernel`.

**Comment by nicola-decao** [comment]
Hi @kshitij12345 !
The function is working and I got the same output as `scipy.special.betainc`:

```python
import torch
import matplotlib.pyplot as plt
from scipy import special

x = torch.rand((100,))
a = torch.rand((100,)) * 10
b = torch.rand((100,)) * 10

from_torch = torch.special.betainc(x, a, b)
from_scipy = special.betainc(a, b, x)

assert torch.isclose(from_torch, from_scipy.float()).all() # works!
```

How do I implement the tests? Is there anything I can look at?

File: aten/src/ATen/native/BinaryOps.cpp [hunk 2]
```diff
@@ -57,6 +57,22 @@ TORCH_META_FUNC(special_xlog1py) (const Tensor& self, const Tensor& other) {
   build_binary_float_op(maybe_get_output(), self, other);
 }
 
+TORCH_META_FUNC(special_betainc) (const Tensor& self, const Tensor& a, const Tensor& b) {
```
Comment (kshitij12345) [inline aten/src/ATen/native/BinaryOps.cpp:60]: TODO: Decide where this should reside!

cc: @mruberry

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 3]
```diff
@@ -973,6 +973,81 @@ void xlog1py_kernel(TensorIteratorBase& iter) {
   });
 }
 
+void betainc_kernel(TensorIteratorBase& iter) {
+  AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, iter.common_dtype(), "betainc_cpu", [&]() {
+    cpu_kernel(iter, [](scalar_t x, scalar_t a, scalar_t b) -> scalar_t {
+      if (at::_isnan(x)){
```
Comment (kshitij12345) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:979]: Would be great if you can factor this lambda into `Math.h` and also stick `C10_HOST_DEVICE` to it, so that we don't have to duplicate this for CPU and CUDA.

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Math.h

You can refer this file to look for `C10_HOST_DEVICE` example (part of WIP PR): 
https://github.com/pytorch/pytorch/blob/29be4bbf32fee23d3e815cfd192ea44ea3857dc1/aten/src/ATen/native/Math.h#L1304

File: tools/autograd/derivatives.yaml [hunk 4]
```diff
@@ -713,6 +713,22 @@
...
+- name: special_betainc.a_scalar(Tensor self, Scalar a, Tensor b) -> Tensor
+  self: not_implemented("special_betainc")
+  b: not_implemented("special_betainc")
+
+- name: special_betainc.b_scalar(Tensor self, Tensor a, Scalar b) -> Tensor
+  self: not_implemented("special_betainc")
+  a: not_implemented("special_betainc")
+
+- name: special_betainc.ab_scalar(Tensor self, Scalar a, Scalar b) -> Tensor
+  self: grad
```
Comment (kshitij12345) [inline tools/autograd/derivatives.yaml:730]: Stray?

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 5]
```diff
@@ -973,6 +973,81 @@ void xlog1py_kernel(TensorIteratorBase& iter) {
...
+      }
+
+      /*Find the first part before the continued fraction.*/
+      const scalar_t lbeta_ab = std::lgamma(a) + std::lgamma(b) - std::lgamma(a + b);
+      const scalar_t front = std::exp(std::log(x) * a + std::log(1.0 - x) * b - lbeta_ab) / a;
+
+      /*Use Lentz's algorithm to evaluate the continued fraction.*/
+      scalar_t f = 1.0, c = 1.0, d = 0.0;
+
+      scalar_t TINY = 1.0e-30, STOP = 1.0e-8, numerator, cd;
```
Comment (kshitij12345) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:1008]: Can TINY and STOP be constexpr

Better to default initialize `numerator` and `cd`

**Review by kshitij12345 — COMMENTED** [review]
Awesome! Things are starting to look good! 

I have added a few comments after a cursory look.
For tests you'll have to do two things,

1. Add an OpInfo entry. For more context look at
     * https://github.com/pytorch/pytorch/wiki/Writing-tests-in-PyTorch-1.8
     * https://github.com/pytorch/pytorch/issues/54261
     * https://github.com/pytorch/pytorch/pull/57428
2. Since OpInfo themselves don't autogenerate the tests for correct implementation, you will need to write a test comparing this implementation against scipy. For reference, https://github.com/kshitij12345/pytorch/blob/c76405d3b1685a4ae0c3fd76d8d69199b340f83a/test/test_binary_ufuncs.py#L2619. Note: You can skip out variants tests as they are tested by OpInfo tests already now.

For 2, I recommend adding them to `test_torch.py` for now. We can decide to move it later.

cc: @mruberry where do you think test against the reference reside?

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 3 above)
Comment (nicola-decao) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:979]: Sure, I just did this and works!

File: tools/autograd/derivatives.yaml (hunk 4 above)
Comment (nicola-decao) [reply tools/autograd/derivatives.yaml:730]: I forgot to put it as `not_implemented`. Fixing now!

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 5 above)
Comment (nicola-decao) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:1008]: If I put them to `constexpr` it gives an error but to `const` it works.

**Comment by mruberry** [comment]
> For 2, I recommend adding them to `test_torch.py` for now. We can decide to move it later.
> 
> cc: @mruberry where do you think test against the reference reside?

Your idea to use test_torch.py sounds great for now. Ultimately we may want to add a test suite for ternary elementwise operations.

**Comment by kshitij12345** [comment]
> Also not sure which file this function should reside. Maybe make a new Special.cpp?
> @mruberry any suggestions?

ping @mruberry 

https://github.com/pytorch/pytorch/pull/58700#issuecomment-845652437

**Comment by mruberry** [comment]
> > Also not sure which file this function should reside. Maybe make a new Special.cpp?
> > @mruberry any suggestions?
> 
> ping @mruberry
> 
> [#58700 (comment)](https://github.com/pytorch/pytorch/pull/58700#issuecomment-845652437)

Thanks for the ping, @kshitij12345. I don't think it matters too much where we put the function, but special.cpp sounds fine.

**Comment by kshitij12345** [comment]
Gentle ping @nicola-decao :). Just wanted to check on the status of this PR.

**Comment by nicola-decao** [comment]
> Gentle ping @nicola-decao :). Just wanted to check on the status of this PR.

Hi @kshitij12345 The function is implemented and works correctly! But I did not implement tests for it and I am not sure how to do it.

**Comment by kshitij12345** [comment]
@nicola-decao Thanks for the update. Unfortunately we are not taking any new operators for special the CUDA context size and the compile timers are not affected by adding these new operators. (You can check the tracking issue description for more reference). Apologies for that.

Once the issue has been sorted. I'll certainly ping you here :)

**Comment by nicola-decao** [comment]
@kshitij12345 I understand! This implementation works on CPU also BTW. I implemented it in a device-agnostic way in `aten/src/ATen/native/Math.h`

**Comment by mruberry** [comment]
Good news! We are now accepting additional special operators with CUDA kernels by just-in-time compiling them. See this note: https://github.com/pytorch/pytorch/blob/35e7ac3fa17bc20d982ea69440d30cd9658dff25/aten/src/ATen/native/cuda/Loops.cuh#L92

And examples in ATen/native/cuda/Math.cuh, like this ndtri defintion: 

https://github.com/pytorch/pytorch/blob/35e7ac3fa17bc20d982ea69440d30cd9658dff25/aten/src/ATen/native/cuda/Math.cuh#L116

It's pretty straightforward but let us know if you have any questions, @nicola-decao!

**Comment by nicola-decao** [comment]
@mruberry @kshitij12345  I do not have time to move on this PR. Since all the code is there and it is working can someone else do it?

Please note that the function is correctly implemented and it returns the same values as `scipy`:
```python
import torch
from scipy import special

x = torch.rand((100,))
a = torch.rand((100,)) * 10
b = torch.rand((100,)) * 10

from_torch = torch.special.betainc(x, a, b)
from_scipy = special.betainc(a, b, x)

assert torch.isclose(from_torch, from_scipy.float()).all() # works!
```

**Comment by 0x00b1** [comment]
@nicola-decao I can pick this up.

**Comment by changebo** [comment]
Thanks for the efforts so far. Are there any updates on this? It would be great if `betainc` is implemented since that will enable computation of the CDF of the student's t distribution. Thanks again!

**Comment by Filco306** [comment]
I would second @changebo 's comment. Is there ay update on this? :) Thank you!!
